interface/ceed-python/ceed_cffi/build-ceed.py

```diff
@@ -0,0 +1,230 @@
+                               CeedTransposeMode tmode,
+                               CeedEvalMode emode, CeedVector u, CeedVector v);
+  int CeedBasisDestroy(CeedBasis *basis);
+
+  int CeedGaussQuadrature(CeedInt Q, CeedScalar *qref1d,
+                                    CeedScalar *qweight1d);
+  int CeedLobattoQuadrature(CeedInt Q, CeedScalar *qref1d,
+                                      CeedScalar *qweight1d);
+  int CeedQRFactorization(Ceed ceed, CeedScalar *mat, CeedScalar *tau,
+                                    CeedInt m, CeedInt n);
+  int CeedSymmetricSchurDecomposition(Ceed ceed, CeedScalar *mat,
+    CeedScalar *lambda, CeedInt n);
+  int CeedSimultaneousDiagonalization(Ceed ceed, CeedScalar *matA,
+    CeedScalar *matB, CeedScalar *x, CeedScalar *lambda, CeedInt n);
+  """ +
+  CeedQFunction +
+  """
+
+typedef int (*CeedQFunctionUser)(void *ctx, const CeedInt Q,
+                                 const CeedScalar *const *in,
+                                 CeedScalar *const *out);
+
+  int CeedQFunctionCreateInterior(Ceed ceed, CeedInt vlength,
+    CeedQFunctionUser f, const char *source, CeedQFunction *qf);
+  int CeedQFunctionCreateInteriorByName(Ceed ceed, const char *name,
+    CeedQFunction *qf);
+  int CeedQFunctionCreateIdentity(Ceed ceed, CeedInt size,
+    CeedQFunction *qf);
+  int CeedQFunctionAddInput(CeedQFunction qf, const char *fieldname,
+                                      CeedInt size, CeedEvalMode emode);
+  int CeedQFunctionAddOutput(CeedQFunction qf, const char *fieldname,
+                                       CeedInt size, CeedEvalMode emode);
+  int CeedQFunctionSetContext(CeedQFunction qf, void *ctx,
+                                        size_t ctxsize);
+  int CeedQFunctionApply(CeedQFunction qf, CeedInt Q,
+                                   CeedVector *u, CeedVector *v);
+  int CeedQFunctionDestroy(CeedQFunction *qf);
+  """ +
+  CeedOperator +
+  """
+  int CeedOperatorCreate(Ceed ceed, CeedQFunction qf,
+                                   CeedQFunction dqf, CeedQFunction dqfT,
+                                   CeedOperator *op);
+  int CeedCompositeOperatorCreate(Ceed ceed, CeedOperator *op);
+  int CeedOperatorSetField(CeedOperator op, const char *fieldname,
+                                     CeedElemRestriction r,
+                                     CeedTransposeMode lmode, CeedBasis b,
+                                     CeedVector v);
+  int CeedCompositeOperatorAddSub(CeedOperator compositeop,
+    CeedOperator subop);
+  int CeedOperatorAssembleLinearQFunction(CeedOperator op,
+    CeedVector *assembled, CeedElemRestriction *rstr, CeedRequest *request);
+  int CeedOperatorAssembleLinearDiagonal(CeedOperator op,
+    CeedVector *assembled, CeedRequest *request);
+  int CeedOperatorApply(CeedOperator op, CeedVector in,
+                                  CeedVector out, CeedRequest *request);
+  int CeedOperatorDestroy(CeedOperator *op);
+  """
+)
```

Comment:
Is there a reason we can't obtain the above by reading `ceed.h`? See `f.read()` patterns at https://cffi.readthedocs.io/en/latest/embedding.html#embedding-and-extending or https://dbader.org/blog/python-cffi

Reply:
Huh. I wonder if that is new(er), as much of the online discussion of cffi states that manually imported the header is(was) required.